### ISSUE TITLE
`Style/ObjectThen` safe autocorrect results in invalid code

### ISSUE THREAD

**Issue by a-lavis** [rubocop/rubocop#12693]
Sometimes the `Style/ObjectThen` cop, which is marked as a safe autocorrect, corrects to invalid ruby code which cannot be parsed.

--------

## Expected behavior

When running `rubocop -a foo.rb` where `foo.rb` is the following file:

```ruby
# frozen_string_literal: true

# rubocop:enable Style/ObjectThen

Foo = Data.define(:input) do
  def one?
    input == 1
  end

  def two?
    input == 2
  end

  def strange_method # rubocop:disable Metrics/MethodLength
    helper = lambda do |foo|
      case foo
      when one?
        puts 'one!'
      when two?
        puts 'two!'
      else
        puts 'other number...'
      end
    end

    yield_self(&helper)
  end
end

Foo.new(3).strange_method

# rubocop:disable Style/ObjectThen
```

I expect `Style/ObjectThen` to not safe autocorrect.

## Actual behavior

It safe autocorrects to:

```ruby
# frozen_string_literal: true

# rubocop:enable Style/ObjectThen

Foo = Data.define(:input) do
  def one?
    input == 1
  end

  def two?
    input == 2
  end

  def strange_method # rubocop:disable Metrics/MethodLength
    helper = lambda do |foo|
      case foo
      when one?
        puts 'one!'
      when two?
        puts 'two!'
      else
        puts 'other number...'
      end
    end

    then(&helper)
  end
end

Foo.new(3).strange_method

# rubocop:disable Style/ObjectThen
```

If you try to run this file by doing `ruby foo.rb` you get the following parse error:
```
> ruby foo.rb
foo.rb: --> foo.rb
unexpected 'then', ignoring itunexpected '&', ignoring itexpected a newline or semicolon after the statementexpected a matching `)`unexpected ')', ignoring it
   5  Foo = Data.define(:input) do
  14    def strange_method # rubocop:disable Metrics/MethodLength
> 26      then(&helper)
  27    end
  28  end

foo.rb:26: syntax error, unexpected `then' (SyntaxError)
    then(&helper)
    ^~~~
```

## Steps to reproduce the problem

Run `rubocop -a foo.rb` where `foo.rb` is the above file.

## RuboCop version

```
> rubocop -V
1.60.2 (using Parser 3.3.0.5, rubocop-ast 1.30.0, running on ruby 3.3.0) [arm64-darwin23]
```
